### PR TITLE
Implement Instruction Selector and Register Bank Selector

### LINKED ISSUE #35: Implement Register Bank Selector
**Is your feature request related to a problem? Please describe.**
Specifically for PIC, it is beneficial to spread data of a statement block in as few data memory banks as possible.

**Describe the solution you'd like**
The Register Bank Selector for GlobalISel may implement allocation of data memory bank.

**Describe alternatives you've considered**
Originally, RegBankSelect is meant for another use case. PIC has just one register bank class which is why RegBankSelect in its original purpose does not really make sense for our use case.  Alternatively, we can 

**Additional context**
- [GlobalISel RegBankSelect](https://llvm.org/docs/GlobalISel/RegBankSelect.html)
- [“Tutorial: GlobalISel”: RegBankSelect](https://youtu.be/Zh4R40ZyJ2k?t=842&si=M9bbO1cgvF9M5Cgc)

### LINKED ISSUE #36: Implement Instruction Select
**Is your feature request related to a problem? Please describe.**
After gMIR has gone through all passes (associated to IRTranslator, Legalizer, and RegBankSelector), the gMIR needs to be turned into actual machine instructions.

**Describe the solution you'd like**
The GlobalISel pass Instruction Select is responsible for selecting instructions given gMIR. We shall implement an Instruction Selector. 

**Describe alternatives you've considered**
We may look at SelectionDAG (deprecated though).

**Additional context**
- [GlobalISel: InstructionSelect](https://llvm.org/docs/GlobalISel/InstructionSelect.html)
- [“Tutorial: GlobalISel”: InstructionSelect](https://youtu.be/Zh4R40ZyJ2k?t=1780&si=POnA2VlhHOoG0bFL)

### PR DESCRIPTION
**Pull request by pr1metine**
Supercedes #49 and fixes #36, fixes #35.

# Register Bank Selector

PIC does not differentiate between integer or float registers. Because of this, all registers including the imaginary ones are all part of the same register bank `any`.

# Instruction Selector

Instruction Selection is mostly written in TableGen. Note that not all instructions are implemented, e.g., control flow. All instructions needed to fulfill the given requirements are implemented.

### CODE REVIEW

**Comment by pr1metine** [comment]
See [these LLVM-PIC requirements](https://github.com/llvm-pic/llvm-pic/wiki/Requirements-🤝)

File: llvm/lib/Target/PICMid/PICMidInstructionSelector.cpp [hunk 1]
```diff
@@ -93,7 +95,41 @@ bool PICMidInstructionSelector::select(MachineInstr &I) {
...
+  switch (I.getOpcode()) {
+  case PICMid::G_SHLE:
+  case PICMid::G_LSHRE:
+    return selectShift(I);
+  default:
+    return false;
+  }
+}
+
+bool PICMidInstructionSelector::selectShift(MachineInstr &I) const {
```
Comment (indicatelovelace) [inline llvm/lib/Target/PICMid/PICMidInstructionSelector.cpp]: Maybe this could be executed earlier.

**Comment by pr1metine** [comment]
With this PR, the following instructions are selected as per `PICMidInstrLogical.td`:

PICMid instruction | Implemented
------------------ | --------------
ADDWF   | 🤝
ANDWF   | 🤝
CLRF    |
CLRW    |
COMF    | 🤝
DECF    | 🤝
DECFSZ  |
INCF    | 🤝
INCFSZ  |
IORWF   | 🤝
MOVF    |
MOVWF   |
NOP     |
RLF     | 🤝
RRF     | 🤝
SUBWF   | 🤝
SWAPF   | 🤝
XORWF   | 🤝
BCF     |
BSF     |
BTFSC   |
BTFSS   |
ADDLW   | 🤝
ANDLW   | 🤝
CALL    |
CLRWDT  |
GOTO    |
IORLW   | 🤝
MOVLW   | 🤝
RETFIE  |
RETLW   | 🤝
RETURN  | 🤝
SLEEP   |
SUBLW   | 🤝
XORLW   | 🤝

File: llvm/test/CodeGen/PICMid/instruction-selector-addsub.mir
Comment (indicatelovelace) [inline llvm/test/CodeGen/PICMid/instruction-selector-addsub.mir:1]: Tests can probaby be autogenerated by the test-suite, See [test-suite][https://llvm.org/docs/TestSuiteGuide.html].

**Review by indicatelovelace — APPROVED** [review]

**Review by handiener — APPROVED** [review]

**Pull request merged**